File: 63-unique-paths-ii/63-unique-paths-ii.cpp
class Solution {
public:
    int func(int i, int j, vector<vector<int>>& grid, vector<vector<int>> &dp){
        if(i == 0 and j == 0) {
            if(grid[i][j] == 0) return 1;
            else return 0;
        }
        else if(i < 0 or j < 0) return 0;
        else if(grid[i][j] == 1) return 0;
        else if(dp[i][j] != -1) return dp[i][j];
        
        //going top down
        int left = func(i, j-1, grid, dp);
        int up = func(i-1, j, grid, dp);
        
        return dp[i][j] = left + up;
    }
    
    int uniquePathsWithObstacles(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        
        vector<vector<int>> dp(m, vector<int>(n, -1));
        
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j] == 1) dp[i][j] = 0;
                else if(i == 0 and j == 0) dp[i][i] = 1;
                else{
                    int up = 0, left = 0;
                    
                    if(i > 0) up = dp[i-1][j];
                    if(j > 0) left = dp[i][j-1];
                    
                    dp[i][j] = left + up;
                }
            }
        }
        return dp[m-1][n-1];
        //return func(m-1, n-1, obstacleGrid, dp);
    }
};

File: 63-unique-paths-ii/63-unique-paths-ii.py
class Solution:
    
    #top down memoized
#         def helper(row, col, grid, dp):
#             if row < 0 or col < 0 or grid[row][col] == 1: return 0
            
#             if row == 0 and col == 0: return 1
#             if (row,col) in dp: return dp[(row, col)]
            
#             left = helper(row, col-1, grid, dp)
#             up = helper(row-1, col, grid, dp)
            
#             dp[(row, col)] = left+up
#             return dp[(row, col)]
    
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        
        #dp = {}       
        
        #bottom up - tabulated
        #dp[(0,0)] = 1
        
        if obstacleGrid[0][0] == 1 or obstacleGrid[-1][-1] == 1: return 0
        
        dp = [[0]* n for _ in range(m)]
        dp[0][0] = 1
        
        
        for i in range(1,m):
            dp[i][0] = 1 if obstacleGrid[i][0]==0 and dp[i-1][0]==1 else 0
            
        for j in range(1,n):
            dp[0][j] = 1 if obstacleGrid[0][j]==0  and dp[0][j-1]==1 else 0
        
        for row in range(1, m):
            for col in range(1, n):
                if obstacleGrid[row][col] != 1: 

                    left = dp[row][col-1]
                    up = dp[row-1][col]

                    dp[row][col] = left+up
        
        
        
        return dp[m-1][n-1]